test/generator/stubtest_jittest.cpp

```diff
@@ -0,0 +1,103 @@
+#include "Halide.h"
+
+#include "stubtest.stub.h"
+
+using Halide::Argument;
+using Halide::Expr;
+using Halide::Func;
+using Halide::Image;
+using Halide::LoopLevel;
+using Halide::Var;
+using StubNS1::StubNS2::StubTest;
+
+const int kSize = 32;
+
+Halide::Var x, y, c;
+
+template<typename Type>
+Image<Type> MakeImage(int extra) {
+    Image<Type> im(kSize, kSize, 3);
+    for (int x = 0; x < kSize; x++) {
+        for (int y = 0; y < kSize; y++) {
+            for (int c = 0; c < 3; c++) {
+                im(x, y, c) = static_cast<Type>(x + y + c + extra);
+            }
+        }
+    }
+    return im;
+}
+
+template<typename InputType, typename OutputType>
+void verify(const Image<InputType> &input, float float_arg, int int_arg, const Image<OutputType> &output) {
+    if (input.width() != output.width() ||
+        input.height() != output.height()) {
+        fprintf(stderr, "size mismatch\n");
+        exit(-1);
+    }
+    int channels = std::max(1, std::min(input.channels(), output.channels()));
+    for (int x = 0; x < output.width(); x++) {
+        for (int y = 0; y < output.height(); y++) {
+            for (int c = 0; c < channels; c++) {
+                const OutputType expected = static_cast<OutputType>(input(x, y, c) * float_arg + int_arg);
+                const OutputType actual = output(x, y, c);
+                if (expected != actual) {
+                    fprintf(stderr, "img[%d, %d, %d] = %f, expected %f\n", x, y, c, (double)actual, (double)expected);
+                    exit(-1);
+                }
+            }
+        }
+    }
+}
+
+int main(int argc, char **argv) {
+    Halide::JITGeneratorContext context(Halide::get_target_from_environment());
+    
+    constexpr int kArrayCount = 2;
+
+    Image<float> src[kArrayCount] = {
+        MakeImage<float>(0),
+        MakeImage<float>(1)
+    };
+
+    std::vector<int> int_args = { 33, 66 };
+
+    // the Stub wants Expr, so make a conversion in place
+    std::vector<Expr> int_args_expr(int_args.begin(), int_args.end());
+
+    // We statically know the types we want, so the templated construction method
+    // is most convenient.
+    auto gen = StubTest::make<>(
+        context, 
+        // Use aggregate-initialization syntax to fill in an Inputs struct.
+        {
+            { Func(src[0]), Func(src[1]) },
+            1.234f, 
+            int_args_expr
+        });
+
+    StubTest::ScheduleParams sp;
+    // This generator defaults intermediate_level to "undefined", 
+    // so we *must* specify something for it (else we'll crater at
+    // Halide compile time). We'll use this:
+    sp.intermediate_level = LoopLevel(gen.f, Var("y"));
```

Comment:
Updating this: as it turns out, the addition of type/dim/size inference to Stubs makes it hard to infer the output Vars... To infer the output Vars for the Stub, we have to call generate() so that all the Output<Func> values are valid. But if *any* Input<> or Output<> values have unspecified values, we can't do that... and we can't just guess at a hopefully-valid value since there may be constraints in the code we can't predict.

I'm pushing a change with a revision to use Func::args() instead:

        sp.intermediate_level = LoopLevel(gen.f, gen.f.args().at(1));

(Note that I'm using args().at() rather than args()[] since the former guarantees exception/halt if entry isn't found, rather than returning a default as with operator[])

After an offline discussion with zalman@, we'd like to hold off on trying to gather the used-by-output-vars into the Stub, mainly because (1) it's hard, and (2) we'd like to actually try out the idiom above in zalman@'s code to see if it feels good enough.

AFAIK this is the last nontrivial stumbling block; LMK your thoughts.